1-Draft/RFCNNNN-Notification-On-Version-Update.md

```diff
@@ -0,0 +1,242 @@
+                // After grab the file lock, it turns out a successful check has finished.
+                // Then let's bail out early.
+                return;
+            }
+
+            // Now it's guaranteed that I'm the only process that reaches here.
+            foreach (string oldFile in Directory.EnumerateFiles(TestDir, "sentinel-*-*-*.done"))
+            {
+                // Clean up the old '.done' file, there should be only one.
+                File.Delete(oldFile);
+            }
+
+            // Do the real update check
+            //  - Send HTTP request to query for the new release/pre-release;
+            //  - If there is a valid new release that should be reported to the user, create the file `_update_<new-version>`,
+            //    or rename the existing `_update_<old-version>` to `_update_<new-version>`.
+            // ... more ...
+
+            // Finally, create the `todayDoneFilePath` file as an indicator that a successful update check has finished today.
+            new FileStream(todayDoneFilePath, FileMode.CreateNew, FileAccess.Write, FileShare.None).Close();
+        }
+    }
+    catch (Exception e)
+    {
+        // An update check is in progress from another `pwsh` process. So it's OK to just return.
+    }
+}
+```
+
+With the file lock, only one process can get in the guarded `using` block at a given time.
+So only one process will be creating the file `_update_<version>_<publish-date>`, or renaming an old such file to reflect the new version.
+Yes, other processes could be looking at the old file name (when a `pwsh` session tries to print a notification),
+or working with an outdated `FileInfo` object (another update check tries to do a pre-validation).
+But it's fine for that to happen:
+
+- In the former case, that particular `pwsh` session will show a notification about an outdated version,
+  but the next `pwsh` session will show the right notification.
+- In the latter case, the other update check will continue and find the `.done` file already exists for today.
+
+It's possible that a `pwsh` session terminates while the update check task is still running,
+in the middle of the `using` block for example.
+Creating the `.done` file is the very last step in the `using` block.
+So if the session ends before the `.done` file is created,
+another update check will happen when the next `pwsh` session starts and finish the work.
+
+#### How to do the update check
+
+This is comparatively the easy part.
+
+- Determine if we need to check pre-releases.
+- Send HTTP query request and parse the response.
+  Some optimization work is needed in this step (see below).
+  It would be much better if we can have the latest release/pre-release information stored in a well-known URL,
+  to make the query easier and take less time.
+  - GitHub API doesn't support querying for the latest pre-release,
+    so we need to hit the 'get-all-releases' API `https://api.github.com/repos/PowerShell/PowerShell/releases`.
```

Comment:
We could query our custom `tools/metadata.json` file at first step. It is more simple and fast.
This allow us to trigger new versions in moderated way (after we check that new packages was published and work well).

Reply:
I don't want this feature to depend on `tools/metadata.json`. That file is an internal artifact for the build and subject to change at any point of time. I don't want to make it a public surface that cannot be changed in future.
@TravisEz13 created a static link `https://aka.ms/pwsh-buildinfo-daily` that points to a `json` file including the daily build tag. Currently there is no equivalent of this for `preview build`, `stable build` and `servicing build`, but he planed to add those too, so every release build will update the json file appropriately and we can query that json file for version information. However, that's not available now.

When that endpoint becomes available, we can update the code to use those endpoints. At the same time, the old code targeting GitHub rest APIs will continue to work.

Reply:
@iSazonov I will work on enabling the static links for `preview` and `stable` builds. We will make it happen before RC.